platform/android/src/snapshotter/map_snapshotter.cpp

```diff
@@ -143,6 +155,134 @@ void MapSnapshotter::deactivateFilesource(JNIEnv& env) {
+    MBGL_VERIFY_THREAD(tid);
+    android::UniqueEnv _env = android::AttachEnv();
+
+    static auto& javaClass = jni::Class<MapSnapshotter>::Singleton(*_env);
+    static auto onDidFinishLoadingStyle = javaClass.GetMethod<void()>(*_env, "onDidFinishLoadingStyle");
+    auto weakReference = javaPeer.get(*_env);
+    if (weakReference) {
+        weakReference.Call(*_env, onDidFinishLoadingStyle);
+    }
+}
+
+void MapSnapshotter::onStyleImageMissing(const std::string& imageName) {
+    MBGL_VERIFY_THREAD(tid);
+    android::UniqueEnv _env = android::AttachEnv();
+    static auto& javaClass = jni::Class<MapSnapshotter>::Singleton(*_env);
+    static auto onStyleImageMissing = javaClass.GetMethod<void(jni::String)>(*_env, "onStyleImageMissing");
+    auto weakReference = javaPeer.get(*_env);
+    if (weakReference) {
+        weakReference.Call(*_env, onStyleImageMissing, jni::Make<jni::String>(*_env, imageName));
+    }
+}
+
+void MapSnapshotter::addLayerAt(JNIEnv& env, jlong nativeLayerPtr, jni::jint index) {
+    assert(nativeLayerPtr != 0);
+    auto layers = snapshotter->getStyle().getLayers();
+    auto* layer = reinterpret_cast<Layer*>(nativeLayerPtr);
+    // Check index
+    int numLayers = layers.size() - 1;
+    if (index > numLayers || index < 0) {
+        Log::Error(Event::JNI, "Index out of range: %i", index);
+        jni::ThrowNew(env,
+                      jni::FindClass(env, "com/mapbox/mapboxsdk/style/layers/CannotAddLayerException"),
+                      std::string("Invalid index").c_str());
+    }
+    // Insert it below the current at that index
+    try {
+        layer->addToStyle(snapshotter->getStyle(), layers.at(index)->getID());
+    } catch (const std::runtime_error& error) {
+        jni::ThrowNew(
+            env, jni::FindClass(env, "com/mapbox/mapboxsdk/style/layers/CannotAddLayerException"), error.what());
+    }
+}
+
+void MapSnapshotter::addLayerBelow(JNIEnv& env, jlong nativeLayerPtr, const jni::String& below) {
+    assert(nativeLayerPtr != 0);
+
+    auto* layer = reinterpret_cast<Layer*>(nativeLayerPtr);
+    try {
+        layer->addToStyle(
+            snapshotter->getStyle(),
+            below ? mbgl::optional<std::string>(jni::Make<std::string>(env, below)) : mbgl::optional<std::string>());
+    } catch (const std::runtime_error& error) {
+        jni::ThrowNew(
+            env, jni::FindClass(env, "com/mapbox/mapboxsdk/style/layers/CannotAddLayerException"), error.what());
+    }
+}
+
+void MapSnapshotter::addLayerAbove(JNIEnv& env, jlong nativeLayerPtr, const jni::String& above) {
+    assert(nativeLayerPtr != 0);
+    auto* layer = reinterpret_cast<Layer*>(nativeLayerPtr);
+
+    // Find the sibling
+    auto layers = snapshotter->getStyle().getLayers();
+    auto siblingId = jni::Make<std::string>(env, above);
+
+    size_t index = 0;
+    for (auto l : layers) {
+        if (l->getID() == siblingId) {
+            break;
+        }
+        index++;
```

Comment:
++index

Reply:
I see everywhere below `index+1` is used, so maybe
```
       ++index;
       if (l->getID() == siblingId) break;
```

Reply:
Good idea.